doc/Configuration.md
# Configuration

The following table lists the configurable parameters of the `Open Vault Agent Injector` chart and their default values.

| Parameter    | Description          | Default                                                         |
|:-------------|:---------------------|:----------------------------------------------------------------|
| image.applicationNameLabel   | Application Name. Must match label com.ovai.application | open-vault-agent-injector   |
| image.metricsPort                | Port exposed for metrics collection | 9000 |
| image.path       | Image path   | asaintsever/open-vault-agent-injector |
| image.port       | Service main port    | 8443            |
| image.pullPolicy   | Pull policy for image: IfNotPresent or Always       | IfNotPresent           |
| image.serviceNameLabel   | Service Name. Must match label com.ovai.service     | open-vault-agent-injector      |
| image.tag  | Image tag     | latest *(local testing)*, [VERSION_OVAI](../VERSION_OVAI) *(release)* |
| imageRegistry  | Image registry |   |
| injectconfig.jobbabysitter.image.path   | Image path | everpeace/curl-jq |
| injectconfig.jobbabysitter.image.pullPolicy | Pull policy for image: IfNotPresent or Always | Always |
| injectconfig.jobbabysitter.image.tag   | Image tag  | latest |
| injectconfig.jobbabysitter.resources.limits.cpu | Job babysitter sidecar CPU resource limits | 120m |
| injectconfig.jobbabysitter.resources.limits.memory | Job babysitter sidecar memory resource limits | 25Mi |
| injectconfig.jobbabysitter.resources.requests.cpu | Job babysitter sidecar CPU resource requests | 100m |
| injectconfig.jobbabysitter.resources.requests.memory | Job babysitter sidecar memory resource requests | 20Mi |
| injectconfig.vault.image.path  | Image path  | vault |
| injectconfig.vault.image.pullPolicy    | Pull policy for image: IfNotPresent or Always  | Always   |
| injectconfig.vault.image.tag  | Image tag | 1.6.5 |
| injectconfig.vault.log.format                    | Vault log format: standard, json    | json    |
| injectconfig.vault.log.level                    | Vault log level: trace, debug, info, warn, err    | info    |
| injectconfig.vault.resources.limits.cpu | Vault Agent CPU resource limits | 200m |
| injectconfig.vault.resources.limits.memory | Vault Agent memory resource limits | 80Mi |
| injectconfig.vault.resources.requests.cpu | Vault Agent CPU resource requests | 160m |
| injectconfig.vault.resources.requests.memory | Vault Agent memory resource requests | 65Mi |
| mutatingwebhook.annotations.appLabelKey | Annotation for application's name. Annotation's value used as Vault role by default. | com.ovai.application  |
| mutatingwebhook.annotations.appServiceLabelKey | Annotation for service's name | com.ovai.service  |
| mutatingwebhook.annotations.keyPrefix | Prefix used for all Open Vault Agent Injector annotations | ovai.asaintsever.org  |
| mutatingwebhook.cert.cacertfile | Default filename for webhook CA certificate (PEM-encoded) in generated or provided Kubernetes Secret | ca.crt |
| mutatingwebhook.cert.certfile | Default filename for webhook certificate (PEM-encoded) in generated or provided Kubernetes Secret | tls.crt |
| mutatingwebhook.cert.certlifetime | Default lifetime in years for generated certificates. Not used if generated is false. | 10 |
| mutatingwebhook.cert.generated | Controls whether webhook certificates, private key and Kubernetes Secret are generated. If not, you have to provide a Kubernetes Secret with name secretName. | true |
| mutatingwebhook.cert.keyfile | Default filename for webhook private key (PEM-encoded) in generated or provided Kubernetes Secret | tls.key |
| mutatingwebhook.cert.secretName | Name of the Kubernetes Secret that contains the webhook certificates and private key. Secret should be in webhook's namespace. To provide if generated is false. | open-vault-agent-injector-cert |
| mutatingwebhook.failurePolicy | Defines how unrecognized errors and timeout errors from the admission webhook are handled. Allowed values are Ignore or Fail | Ignore |
| mutatingwebhook.loglevel | Enable V-leveled logging at the specified level | 4 |
| mutatingwebhook.namespaceSelector.boolean    | Enable to control, with label "vault-injection=enabled", the namespaces where injection is allowed (if false: all namespaces except _kube-system_ and _kube-public_) | false                                                           |
| mutatingwebhook.namespaceSelector.namespaced | Enable to control, with label "vault-injection={{ .Release.Namespace }}", the specific namespace where injection is allowed (ie, restrict to namespace where injector is installed) | false |
| probes.liveness.failureThreshold                | Number of probe failure before restarting the probe                                 | 3  |
| probes.liveness.initialDelaySeconds             | Number of seconds after the container has started before the probe is initiated     | 2  |
| probes.liveness.periodSeconds                   | How often (in seconds) to perform the probe                                         | 20 |
| probes.liveness.timeoutSeconds                  | Number of seconds after which the probe times out                                   | 5  |
| probes.readiness.failureThreshold               | Number of probe failure before setting the probe to Unready                         | 3  |
| probes.readiness.initialDelaySeconds            | Number of seconds after the container has started before the probe is initiated     | 2  |
| probes.readiness.periodSeconds                  | How often (in seconds) to perform the probe       | 20   |
| probes.readiness.successThreshold      | Minimum consecutive successes for the probe to be considered successful after having failed  | 1  |
| probes.readiness.timeoutSecon          | Number of seconds after which the probe times out  | 5   |
| registryKey         | Name of Kubernetes secret for image registry                        |  |
| replicaCount                        | Number of replicas | 3    |
| resources.limits.cpu                | CPU resource limits                             | 250m         |
| resources.limits.memory             | Memory resource limits                          | 256Mi        |
| resources.requests.cpu              | CPU resource requests                           | 100m         |
| resources.requests.memory           | Memory resource requests                        | 128Mi        |
| revisionHistoryLimit                | Revision history limit in tiller / helm / k8s   | 3            |
| service.exposedServicePort   | Port exposed by the K8s service (Kubernetes always assumes port 443 for webhooks) | 443 |
| service.name                                    | Service name            | open-vault-agent-injector                                   |
| service.prefixWithHelmRelease                   | Service name to be prefixed with Helm release name                                       | false                                                           |
| service.type                        | Kubernetes service type: ClusterIP, NodePort, LoadBalancer, ExternalName  | ClusterIP    |
| vault.addr                          | Address of Vault server    | `null` - To be provided at deployment time (e.g.: https://vault:8200)   |
| vault.authMethods.approle.path      | Path defined for AppRole Auth Method            | approle |
| vault.authMethods.approle.roleid_filename    | Filename for role id    | approle_roleid   |
| vault.authMethods.approle.secretid_filename  | Filename for secret id  | approle_secretid |
| vault.authMethods.kubernetes.path      | Path defined for Kubernetes Auth Method            | kubernetes |
| vault.ssl.verify               | Enable or disable verification of certificates               | true |

You can override these values at runtime using the `--set key=value[,key=value]` argument to `helm install`. For example,

```bash
$ helm install open-vault-agent-injector \
               <chart_folder_location> \
               --namespace <your namespace> \
               --set <parameter1>=<value1>,<parameter2>=<value2>
```


CHANGELOG.md
# Changelog for Open Vault Agent Injector

## Release v1.0.0 - 2021-09-21

**Changed**

- [OVAI #1](https://github.com/asaintsever/open-vault-agent-injector/pull/1) - First OVAI release (*forked from Talend's [Vault Sidecar Injector](https://github.com/Talend/vault-sidecar-injector/releases/tag/v7.2.1) v7.2.1 release*)

<hr/>

> **Note on Talend's Vault Sidecar Injector releases**
>
> Open Vault Agent Injector *(OVAI)* is an open source fork of Talend's Vault Sidecar Injector repository, based on the **[v7.2.1 release](https://github.com/Talend/vault-sidecar-injector/releases/tag/v7.2.1)**. Starting from this release, ***OVAI content may diverge and there is no guarantee both projects remain in sync***.
>
> You can look at the CHANGELOG for details on all fixes and features of Talend's Vault Sidecar Injector [here](https://github.com/Talend/vault-sidecar-injector/blob/master/CHANGELOG.md).


cmd/vaultinjector-webhook/flags.go
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"flag"
	"fmt"
	"os"

	"k8s.io/klog"
)

func parseFlags() string {
	// Cert command parameters
	certCmd := flag.NewFlagSet(CertCmd, flag.ExitOnError)
	certCmd.StringVar(&certParameters.CertOperation, "certop", CreateCert, "operation on webhook certificates (create, delete)")
	certCmd.StringVar(&certParameters.CertSecretName, "certsecretname", "", "name of generated or provided Kubernetes secret storing webhook certificates and private key")
	certCmd.StringVar(&certParameters.CertHostnames, "certhostnames", "", "host names to register in webhook certificate (comma-separated list)")
	certCmd.IntVar(&certParameters.CertLifetime, "certlifetime", 10, "lifetime in years for generated certificates")
	certCmd.StringVar(&certParameters.CACertFile, "cacertfile", "ca.crt", "default filename for webhook CA certificate (PEM-encoded) in generated or provided k8s secret")
	certCmd.StringVar(&certParameters.CertFile, "certfile", "tls.crt", "default filename for webhook certificate (PEM-encoded) in generated or provided k8s secret")
	certCmd.StringVar(&certParameters.KeyFile, "keyfile", "tls.key", "default filename for webhook private key (PEM-encoded) in generated or provided k8s secret")

	// Webhook command parameters
	webhookCmd := flag.NewFlagSet(WebhookCmd, flag.ExitOnError)
	webhookCmd.IntVar(&webhookParameters.Port, "port", 8443, "webhook server port")
	webhookCmd.IntVar(&webhookParameters.MetricsPort, "metricsport", 9000, "metrics server port (Prometheus)")
	webhookCmd.StringVar(&webhookParameters.CACertFile, "cacertfile", "ca.crt", "PEM-encoded webhook CA certificate")
	webhookCmd.StringVar(&webhookParameters.CertFile, "certfile", "tls.crt", "PEM-encoded webhook certificate used for TLS")
	webhookCmd.StringVar(&webhookParameters.KeyFile, "keyfile", "tls.key", "PEM-encoded webhook private key used for TLS")
	webhookCmd.StringVar(&webhookParameters.WebhookCfgName, "webhookcfgname", "", "name of MutatingWebhookConfiguration resource")
	webhookCmd.StringVar(&webhookParameters.AnnotationKeyPrefix, "annotationkeyprefix", "ovai", "annotations key prefix")
	webhookCmd.StringVar(&webhookParameters.AppLabelKey, "applabelkey", "application.name", "key for application label")
	webhookCmd.StringVar(&webhookParameters.AppServiceLabelKey, "appservicelabelkey", "service.name", "key for application's service label")
	webhookCmd.StringVar(&webhookParameters.InjectionCfgFile, "injectioncfgfile", "", "file containing the mutation configuration (initcontainers, sidecars, volumes, ...)")
	webhookCmd.StringVar(&webhookParameters.ProxyCfgFile, "proxycfgfile", "", "file containing Vault proxy configuration")
	webhookCmd.StringVar(&webhookParameters.TemplateBlockFile, "tmplblockfile", "", "file containing the template block")
	webhookCmd.StringVar(&webhookParameters.TemplateDefaultFile, "tmpldefaultfile", "", "file containing the default template")
	webhookCmd.StringVar(&webhookParameters.PodLifecycleHooksFile, "podlchooksfile", "", "file containing the lifecycle hooks to inject in the requesting pod")

	if len(os.Args) == 1 {
		usage(os.Args[0])
		os.Exit(1)
	}

	viSubcommand := os.Args[1]

	switch viSubcommand {
	case CertCmd:
		klog.InitFlags(certCmd)
		certCmd.Parse(os.Args[2:])
	case WebhookCmd:
		klog.InitFlags(webhookCmd)
		webhookCmd.Parse(os.Args[2:])
	case VersionCmd:
		fmt.Println("OVAI (Open Vault Agent Injector) version " + VERSION)
		os.Exit(0)
	default:
		usage(os.Args[0])
		os.Exit(1)
	}

	return viSubcommand
}

func usage(program string) {
	fmt.Printf("Usage: %s <command> [<args>]\n\nCommands:\n", program)
	fmt.Printf("  %s\n", CertCmd)
	fmt.Printf("  %s\n", WebhookCmd)
	fmt.Printf("  %s\n", VersionCmd)
	fmt.Printf("\nUse \"%s <command> --help\" for more information about a given command.\n", program)
}
